optimize CompareSameName

arifogel opened the pull request:
`CompareSameName` now runs asymptotically faster by hashing and caching JSON representation of structures.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/batfish/batfish/787)
<!-- Reviewable:end -->

dhalperi commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p7Q82zW_XfEkxCqUr:-L0p7Q82zW_XfEkxCqUs:b-b8q80y) (no related file):*
Worth adding a reasonably small number of unit tests, I think.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 29 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p3b850921hvpWv1Hx:-L0p3b850921hvpWv1Hy:b-xl05lr) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L29)):*
> ```Java
>   }
> 
>   public NamedStructureEquivalenceSet(T namedStructure, SortedSet<String> nodes) {
> ```

Swap argument order to keep it consistent with the previous constructor?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 31 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p3Tx9ThwPapiZHNe0:-L0p3Tx9ThwPapiZHNe1:b-y7grio) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L31)):*
> ```Java
>   public NamedStructureEquivalenceSet(T namedStructure, SortedSet<String> nodes) {
>     _namedStructure = namedStructure;
>     _nodes = nodes;
> ```

should this be `ImmutableSortedSet.copyOf(nodes)`?

(which will be a no-op if nodes is already immutable.)

If so, should the input also be an `Iterable<String>` to accept more use cases?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 71 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p3vBkI3rOyOssD8tV:-L0p3vBkI3rOyOssD8tW:b-qinw9m) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L71)):*
> ```Java
>     }
>     sb.append("\n");
>     return sb.toString();
> ```

This function can now be simplified to `return String.format("%s%s\n", indent, String.join(" ", _nodes));`, right?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 83 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p4Hj_Lo-g9oJ3syuL:-L0p4Hj_Lo-g9oJ3syuM:b-qmxzmm) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L83)):*
> ```Java
> 
>   @JsonProperty(PROP_REPRESENTATIVE_ELEMENT)
>   private void setRepresentativeElement(String representativeElement) {}
> ```

Add a comment for why it's okay to ignore this JSON property? I guess it should come in via `nodes` instead?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSets.java, line 29 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p4xxcuVKePXds2nOO:-L0p4xxcuVKePXds2nOP:b-682k3g) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSets.java#L29)):*
> JSONAssert.assertEquals

Do you not want to use `JSONCompare.compareJSON` and avoid the try/catch?

From the code for JSONAssert, looks like you would be using LENIENT compare mode:

```java
    /**
     * Asserts that the JSONArray provided matches the expected string.  If it isn't it throws an
     * {@link AssertionError}.
     *
     * @param expectedStr Expected JSON string
     * @param actualStr String to compare
     * @param strict Enables strict checking
     * @throws JSONException
     */
    public static void assertEquals(String expectedStr, String actualStr, boolean strict)
            throws JSONException {
        assertEquals(expectedStr, actualStr, strict ? JSONCompareMode.STRICT : JSONCompareMode.LENIENT);
    }
```

---

*[projects/question/src/main/java/org/batfish/question/CompareSameNameQuestionPlugin.java, line 203 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p6Az6ivX5NYeq12ax:-L0p6Az6ivX5NYeq12ay:b-8lrarh) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/question/src/main/java/org/batfish/question/CompareSameNameQuestionPlugin.java#L203)):*
> ```Java
>         for (String structName : allNames) {
>           if (structName.startsWith("~")) {
>             continue;
> ```

I think you can delete this filter now?

Actually, it looks like you just didn't quite finish combining these two loops, but I could be wrong. It looks all this logic could be combined to produce a single multimap.

---

*[projects/question/src/main/java/org/batfish/question/OutliersQuestionPlugin.java, line 164 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p7F5D3ldXryuWFddd:-L0p7F5D3ldXryuWFdde:b-9ha7bk) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/question/src/main/java/org/batfish/question/OutliersQuestionPlugin.java#L164)):*
> ```Java
>             eClasses
>                 .stream()
>                 .filter(eClass -> eClass != max)
> ```

is this intentionally `!=` instead of `!.equals()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/787)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 3 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p7Q82zW_XfEkxCqUr:-L0pJ8GLOQmCayxGTW9Z:b-sm7b0n) (no related file):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Worth adding a reasonably small number of unit tests, I think.
</blockquote></details>

Filed #792 as test tax.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 29 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p3b850921hvpWv1Hx:-L0pDZe1X14du2z2osZR:bb6un1j) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L29)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Swap argument order to keep it consistent with the previous constructor?
</blockquote></details>

Done

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 31 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p3Tx9ThwPapiZHNe0:-L0pDgnTD0Dt0Uo9LoHG:b-wyasdm) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L31)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

should this be `ImmutableSortedSet.copyOf(nodes)`?

(which will be a no-op if nodes is already immutable.)

If so, should the input also be an `Iterable<String>` to accept more use cases?
</blockquote></details>

Sure OK. Done

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 71 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p3vBkI3rOyOssD8tV:-L0pE58Z9VbVC5azZLZ9:b-yia8qb) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L71)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

This function can now be simplified to `return String.format("%s%s\n", indent, String.join(" ", _nodes));`, right?
</blockquote></details>

Due to sorting, yes.
Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java, line 83 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p4Hj_Lo-g9oJ3syuL:-L0pEOlqazyPus788mvD:b-cbbhim) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSet.java#L83)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Add a comment for why it's okay to ignore this JSON property? I guess it should come in via `nodes` instead?
</blockquote></details>

Exactly.
Added comment.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSets.java, line 29 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p4xxcuVKePXds2nOO:-L0pFCBzgleL8mj6WY0S:bqn4jst) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/collections/NamedStructureEquivalenceSets.java#L29)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> JSONAssert.assertEquals

Do you not want to use `JSONCompare.compareJSON` and avoid the try/catch?

From the code for JSONAssert, looks like you would be using LENIENT compare mode:

```java
    /**
     * Asserts that the JSONArray provided matches the expected string.  If it isn't it throws an
     * {@link AssertionError}.
     *
     * @param expectedStr Expected JSON string
     * @param actualStr String to compare
     * @param strict Enables strict checking
     * @throws JSONException
     */
    public static void assertEquals(String expectedStr, String actualStr, boolean strict)
            throws JSONException {
        assertEquals(expectedStr, actualStr, strict ? JSONCompareMode.STRICT : JSONCompareMode.LENIENT);
    }
```
</blockquote></details>

I don't see the connection between this Javadoc and avoiding a try/catch. In any case I really don't want to mess with this right now. This is what we were doing before.

---

*[projects/question/src/main/java/org/batfish/question/CompareSameNameQuestionPlugin.java, line 203 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p6Az6ivX5NYeq12ax:-L0pFTrwXIHCeMjdRPAg:bvuywap) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/question/src/main/java/org/batfish/question/CompareSameNameQuestionPlugin.java#L203)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

I think you can delete this filter now?

Actually, it looks like you just didn't quite finish combining these two loops, but I could be wrong. It looks all this logic could be combined to produce a single multimap.
</blockquote></details>

In fact there is a serious problem here, which is that we are not going to compare generated route-map clauses (in case of NX-OS `continue` llnes).
For now I just removed the extraneous filter.
I filed #791 as a cleanup issue.

---

*[projects/question/src/main/java/org/batfish/question/OutliersQuestionPlugin.java, line 164 at r1](https://reviewable.io:443/reviews/batfish/batfish/787#-L0p7F5D3ldXryuWFddd:-L0pIdQmRarE6s77sMKF:bkmo1fr) ([raw file](https://github.com/batfish/batfish/blob/b35a47dd288b1c968fb7f82f4b279b266eecdf1e/projects/question/src/main/java/org/batfish/question/OutliersQuestionPlugin.java#L164)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

is this intentionally `!=` instead of `!.equals()`?
</blockquote></details>

Yes. Though in this case it makes no difference, as `NamedStructureEquivalenceSet<T>` does not override `equals` nor `hashCode`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/787)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/787#-:-L0pVd5M5BSv6xev4nyO:bnfp4nl)*
<!-- Sent from Reviewable.io -->
